wrapper/hosts.py

```diff
@@ -595,77 +602,289 @@ def sdk_connection(self, data):
-                                          disk_id)
-                # Avoid checking timeouts, and waiting, if there are no
-                # more disks to remove
-                if len(disk_ids) > 0:
-                    if time.time() > endt:
-                        logging.error('Timed out waiting for disks: %r',
-                                      disk_ids)
-                        break
-                    time.sleep(1)
+            system_service = conn.system_service()
+            # We could probably just clean up everything no matter whether it's
+            # two phase or not.
+            if data['two_phase']:
+                self._handle_twophase_cleanup(system_service, data)
+            else:
+                self._handle_simple_cleanup(system_service, data)
+
+    def handle_finish(self, data):
+        if data['two_phase']:
+            with self.sdk_connection(data) as conn:
+                system_service = conn.system_service()
+                self._detach_disks(system_service, self._conversion_vm)
+                self._update_disks(conn)
+
+    def _handle_twophase_cleanup(self, system_service, data):
+        self._detach_disks(system_service, self._conversion_vm)
+        if STATE.vm_id:
+            vm = system_service.vms_service().vm_service(STATE.vm_id)
+            self._detach_disks(system_service, vm.get())
+            vm.remove()
+        self._remove_disks(system_service)
+
+    def _handle_simple_cleanup(self, system_service, data):
+        transfers_service = system_service.image_transfers_service()
+        disk_ids = list(STATE.internal['disk_ids'].values())
+        # First stop all active transfers...
+        try:
+            transfers = transfers_service.list()
+            transfers = [t for t in transfers if t.image.id in disk_ids]
+            if len(transfers) == 0:
+                logging.debug('No active transfers to cancel')
+            for transfer in transfers:
+                logging.info('Canceling transfer id=%s for disk=%s',
+                             transfer.id, transfer.image.id)
+                transfer_service = \
+                    transfers_service.image_transfer_service(
+                        transfer.id)
+                transfer_service.cancel()
+                # The incomplete disk will be removed automatically
+                disk_ids.remove(transfer.image.id)
+        except self.sdk.Error:
+            logging.exception('Failed to cancel transfers')
+
+        # ... then delete the uploaded disks
+        self._delete_disks(system_service.disks_service(), disk_ids)
+
+    def _delete_disks(self, disks_service, disk_ids):
+        logging.info('Removing disks: %r', disk_ids)
+        endt = time.time() + TIMEOUT
+        while len(disk_ids) > 0:
+            for disk_id in disk_ids[:]:
+                disk_service = disks_service.disk_service(str(disk_id))
+                disk = disk_service.get()
```

Comment:
I moved this because I was quite sure it works This is a move of pre-existing code. Also the IDs are only added to the list when they are detected from the virt-v2v output. But I will put it back as requested.